trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/RequestFilter.java

```diff
@@ -0,0 +1,198 @@
+            try {
+                signingMetadata = signingController.validateAndParseAuthorization(request, signingServiceType);
+            }
+            catch (Exception e) {
+                requestLoggingSession.logException(e);
+
+                try {
+                    switch (Throwables.getRootCause(e)) {
+                        case WebApplicationException webApplicationException -> throw webApplicationException;
+                        case IOException ioException -> throw ioException;
+                        default -> throw new RuntimeException(e);
+                    }
+                }
+                finally {
+                    requestLoggingSession.close();
+                }
+            }
+
+            containerRequest.setProperty(RequestFilter.class.getName(), new InternalRequestContext(request, signingMetadata, requestLoggingSession));
+        }
+        else {
+            log.warn("%s is not a ContainerRequest", requestContext.getRequest().getClass().getName());
+            throw new WebApplicationException(BAD_REQUEST);
+        }
+    }
+
+    @Override
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext)
+            throws IOException
+    {
+        if ((requestContext.getRequest() instanceof ContainerRequest containerRequest) && (responseContext instanceof ContainerResponse containerResponse)) {
+            Optional.ofNullable(unwrapContext(containerRequest))
+                    .ifPresent(internalRequestContext -> {
+                        OutputStream entityStream = (containerResponse.isCommitted() || !containerResponse.hasEntity()) ? null : responseContext.getEntityStream();
+                        if (entityStream != null) {
+                            responseContext.setEntityStream(closingLoggerStream(internalRequestContext.requestLoggingSession, entityStream));
+                        }
+                        else {
+                            internalRequestContext.requestLoggingSession.close();
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public Function<ContainerRequest, ?> getValueProvider(Parameter parameter)
+    {
+        if (Request.class.isAssignableFrom(parameter.getRawType())) {
+            return containerRequest -> unwrapContext(containerRequest).request;
+        }
+        if (SigningMetadata.class.isAssignableFrom(parameter.getRawType())) {
+            return containerRequest -> unwrapContext(containerRequest).signingMetadata;
+        }
+        if (RequestLoggingSession.class.isAssignableFrom(parameter.getRawType())) {
+            return containerRequest -> unwrapContext(containerRequest).requestLoggingSession;
+        }
+        return null;
+    }
+
+    @Override
+    public PriorityType getPriority()
+    {
+        return HIGH;
+    }
+
+    private static OutputStream closingLoggerStream(RequestLoggingSession requestLoggingSession, OutputStream delegate)
```

Comment:
This can be more generic, it doesn't care use any `RequestLoggingSession` specific functions
```suggestion
    private static OutputStream closingStream(Closable closable, OutputStream delegate)
```